assignment2/q2.cpp

```diff
@@ -0,0 +1,148 @@
+#include <iostream>
+using namespace std;
+
+struct Node{
+    int value;
+    Node *parent;
+    Node *left, *right;
+};
+
+bool FindNode(Node *root, Node *to_find) {
+
+    if(root == NULL)
+        return false;
+
+    if(root == to_find)
+        return true;
+
+    if(FindNode(root->left, to_find) || FindNode(root->right, to_find))
+        return true;
+
+    return false;
+}
+
+//O(N^2) without any additional memory or changing structure
+//could be O(N) if i.e. it stores paths to nodes and then compares them
+Node *FindCommonAncestor(Node *root, Node *first, Node *second) {
+
+    if(root == NULL)
+        return NULL;
+
+    //nodes are not in the tree
+    if(!FindNode(root, first) || !FindNode(root, second))
```

Comment:
Idea for optimization (not mandatory):
You can simulate the solution of `FindNode` in `root` by using the results from `FindNode` in the left and right children, and checking the root node itself.
This will save you searching the tree 4 times each call (instead of only 2, once for each first/second node)